src/diffusers/models/attention.py

```diff
@@ -92,6 +92,20 @@ def forward(self, x: torch.Tensor, objs: torch.Tensor) -> torch.Tensor:
         return x
 
 
+class TemporalSelfAttention(Attention):
```

Comment:
Is this just applying the sinusoidal position embedding before passing through attention? 

If that's the case then the `BasicTransformerBlock` can handle it. Set `positional_embeddings="sinusoidal` and `num_positional_embeddings= temporal_position_encoding_max_len`

Reply:
From loading the state dict and going through the reference animatediff-xl code, I see that every Attention layer (attn1, attn2) has a pos_embed attribute.

<details>
<summary> Small part of motion sdxl state dict </summary>

```py
MotionAdapterXL(
  (down_blocks): ModuleList(
    (0): MotionModules(
      (motion_modules): ModuleList(
        (0-1): 2 x TransformerTemporalModel(
          (norm): GroupNorm(32, 320, eps=1e-06, affine=True)
          (proj_in): Linear(in_features=320, out_features=320, bias=True)
          (transformer_blocks): ModuleList(
            (0): BasicTransformerBlock(
              (norm1): LayerNorm((320,), eps=1e-05, elementwise_affine=True)
              (attn1): TemporalSelfAttention(
                (to_q): LoRACompatibleLinear(in_features=320, out_features=320, bias=False)
                (to_k): LoRACompatibleLinear(in_features=320, out_features=320, bias=False)
                (to_v): LoRACompatibleLinear(in_features=320, out_features=320, bias=False)
                (to_out): ModuleList(
                  (0): LoRACompatibleLinear(in_features=320, out_features=320, bias=True)
                  (1): Dropout(p=0.0, inplace=False)
                )
                (pos_embed): SinusoidalPositionalEmbedding()
              )
              (norm2): LayerNorm((320,), eps=1e-05, elementwise_affine=True)
              (attn2): TemporalSelfAttention(
                (to_q): LoRACompatibleLinear(in_features=320, out_features=320, bias=False)
                (to_k): LoRACompatibleLinear(in_features=320, out_features=320, bias=False)
                (to_v): LoRACompatibleLinear(in_features=320, out_features=320, bias=False)
                (to_out): ModuleList(
                  (0): LoRACompatibleLinear(in_features=320, out_features=320, bias=True)
                  (1): Dropout(p=0.0, inplace=False)
                )
                (pos_embed): SinusoidalPositionalEmbedding()
              )
              (norm3): LayerNorm((320,), eps=1e-05, elementwise_affine=True)
              (ff): FeedForward(
                (net): ModuleList(
                  (0): GEGLU(
                    (proj): LoRACompatibleLinear(in_features=320, out_features=2560, bias=True)
                  )
                  (1): Dropout(p=0.0, inplace=False)
                  (2): LoRACompatibleLinear(in_features=1280, out_features=320, bias=True)
                )
              )
            )
          )
          (proj_out): Linear(in_features=320, out_features=320, bias=True)
        )
      )
    )
```

</details>

If I set `positional_embeddings="sinusoidal"`, the `pos_embed` layer is in the BasicTransformerBlock itself instead of the individual attention layers (like above), which leads to key mismatch and error while loading.

https://github.com/huggingface/diffusers/blob/bf40d7d82a732a35e0f5b907e59064ee080cde9f/src/diffusers/models/attention.py#L171

We basically want these pos_embed layers to be in every attention layer for the keys to match perfectly. Let me know if I'm doing this wrong :sweat_smile:

Reply:
If these `pos_embed`'s are the exact same thing for each attention layer and that's the reason why it has been made a part of BasicTransformerBlock, I can try and pop the following keys from motion state dict:

- `down_blocks.{N}.motion_modules.{M}.transformer_blocks.{K}.attn1.pos_embed`
- `down_blocks.{N}.motion_modules.{M}.transformer_blocks.{K}.attn2.pos_embed`

and replace it with a single copy in the BasicTransformerBlock itself.

Reply:
Turns out they are the same indeed :sweat_smile:

```torch.all(torch.eq(ckpt_state_dict['down_blocks.0.motion_modules.0.transformer_blocks.0.attn1.pos_embed.pe'], ckpt_state_dict['down_blocks.0.motion_modules.0.transformer_blocks.0.attn2.pos_embed.pe']))```

results in `tensor(True)` for all such occurences. Will refactor it back to just use Attention.